esp-hal/src/i2s/master.rs

```diff
@@ -251,6 +315,429 @@ impl DataFormat {
+    #[default]
+    HalfFrame,
+    /// World select signal will be kept active for the length of the first channel (PCM long frame
+    /// standard)
+    #[cfg(not(any(esp32, esp32s2)))]
+    OneChannel,
+    /// World select signal will be kept active for a single BCLK cycle (PCM short frame standard)
+    Bit,
+    /// World select signal will be kept active for the specified amount of bits(BCLK cycles)
+    #[cfg(not(any(esp32, esp32s2)))]
+    Bits(u16),
+}
+
+/// Represents the polarity of a signal
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Polarity {
+    /// The signal is high when active
+    #[default]
+    ActiveHigh,
+    /// The signal is low when active
+    ActiveLow,
+}
+
+/// I2S channels configuration
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct Channels {
+    count: u8,
+    mask: u16,
+    fill: Option<u32>,
+}
+
+impl Channels {
+    /// Two channels will use different data
+    pub const STEREO: Channels = Channels::new_impl(2, 0b11, None);
+    /// Two channels will use the same data
+    pub const MONO: Channels = Channels::new_impl(2, 0b01, None);
+    /// Two channels. Left(first) channel will contain data. Right(second) channel will contain
+    /// zeros.
+    pub const LEFT: Channels = Channels::new_impl(2, 0b01, Some(0));
+    /// Two channels. Right(second) channel will contain data. Left(first) channel will contain
+    /// zeros.
+    pub const RIGHT: Channels = Channels::new_impl(2, 0b10, Some(0));
+
+    #[procmacros::doc_replace]
+    /// Creates arbitrary configuration for I2S channels.
+    ///
+    /// - `count` the total number of channels. Must be at least 1 and no more than 16.
+    /// - `mask` determines which channels will be active, with the least significant bit
+    ///   representing first channel. Setting the bit at the nth position means nth channel is
+    ///   active. Inactive channels do not consume or write data in the DMA buffer.
+    /// - `fill` determines the behavior of inactive channels. `Some(n)` will make all inactive
+    ///   channel send out specified value, truncated to the channel width. `None` will make
+    ///   disabled channels repeat the data from the last active channel. This field is ignored in
+    ///   the receiver unit.
+    ///
+    /// ## Example
+    ///
+    /// The following example prepares configuration for 6 channels. Only 1st and 4th channels
+    /// are active. Channels 2-3 will use the same data as the 1st, and channels 5-6 will use the
+    /// data from the 4th.
+    ///
+    /// ```rust, no_run
+    /// # {before_snippet}
+    /// use esp_hal::i2s::master::Channels;
+    ///
+    /// let channels = Channels::new(6, 0b_001_001, None);
+    /// # {after_snippet}
+    /// ```
+    #[cfg(any(esp32c3, esp32s3))]
+    pub const fn new(count: u8, mask: u16, fill: Option<u32>) -> Self {
```

Comment:
Why is this public if `configure` just panics if a value other than the 4 presets are passed?